### ISSUE TITLE
Add ability to restrict overriding of terms with graph containers

### ISSUE THREAD

**Issue by BigBlueHat** [w3c/json-ld-syntax#87]
Related to #20 -- or at least proposed off the cuff by yours truly. 😁

**Comment by BigBlueHat** [comment]
This approach would also potentially relate to scoped contexts.

**Comment by azaroth42** [comment]
From call - Allow scoped contexts on graph containers in some form?

**Comment by BigBlueHat** [comment]
Graph containers seem to map to @dlongley's desired feature of "sealing a term, but not its contents."

The idea was (iirc) to essentially limit the ability to extend a document to things within a graph container--such that a sealed *context* would still be extensible within a graph container.

So...
```json
{
  "@context": "https://example.com/sealed-context.jsonld",
  "name": "Term is sealed",
  "items": [
    {"@context": "https://example.com/extended-context.jsonld",
     "name": "Not the same as the one above"}
  }
}
```

Conversations have moved forward sense, then, but I'd mainly been trying to find a "natural" place for context extensibility where the areas of extensibility were clear--rather than "random" terms being sealed upstream by which ever context happened to get there first.

However, this clearly limits the extensibility of a JSON document to what can be expressed *under* single terms--rather than adding extensions to the top-level keys...but that has it's own advantages. 😃 

Essentially, it sees the "parent" document as a sealed envelop for other "child" documents (within the graph container) to carry their own semantics within the contents of that envelop.

**Comment by gkellogg** [comment]
Sealed graphs containers fall out of the API design, but to get the desired behavior, "items" would be sealed, but would have a scoped context of `null`:

```json
{
  "@context": {
    ...
    "name": ...,
    "items": {..., "@context": null}
  }
}
```

This would really only be necessary if it was thought that extended-context.jsonld would re-define terms from sealed-context.jsonld, but IIRC, there is no notion of active context intended underneath `items` in any case.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-02-07-json-ld#section2).

- `RESOLVED:  Adopt the following rules for sealing - 1: If @sealed:true is encountered on a term definition in the active context, then attempts to redefine the term will fail and result in a warning. 2 - If @sealed:true is encountered on a context, then all terms defined within the context should be treated as having @sealed:true in their definition. 3- If @sealed:false is encountered on a term definition in the active context, where the context has been sealed with @sealed:true, then the sealing for that term is removed. {: #resolution1 .resolution}`
- `RESOLVED:  Close #87, as unnecessary given above rules. {: #resolution2 .resolution}`
- `RESOLVED:  Close #98, as @context: null clears context definitions {: #resolution3 .resolution}`
- `RESOLVED:  Close #98, as @context: null clears context definitions {: #resolution4 .resolution}`
- `RESOLVED:  if @id is not present in a term definition, and there is no default vocabulary, but has an expansion to an absolute IRI in the active context, then the @id of the new definition is taken from the active context (otherwise this is still an error) {: #resolution5 .resolution}`
- `RESOLVED:  Once previous is done, we can close #116 {: #resolution6 .resolution}`
- `RESOLVED:  We agree with the processing order per #61: scoped definition from property, then from type, then embedded. PR to update docs to come. {: #resolution7 .resolution}`
<details><summary><i class="differentiate">View the transcript</i></summary>
Sealed contexts<br/>
Ivan Herman: If we can’t get to a clear model and agreement on sealing today, I propose we defer it.<br/>
Adam Soroka: If we happen to do that, what do we say to verifiable claims, etc?<br/>
Rob Sanderson: (reads guiding principles)<br/>
Rob Sanderson: <a href="https://www.w3.org/2018/json-ld-wg/WorkMode/guiding_principles">https://www.w3.org/2018/json-ld-wg/WorkMode/guiding_principles</a><br/>
Rob Sanderson: sealed contexts - goal is to solve by end of day<br/>
Ivan Herman: one thing from last week’s discussions - sealing seems to be a very dangerous thing to do…<br/>
… if Google decides to put a seal on schema.org, community will be unhappy …<br/>
Gregg Kellogg: we need to understand who we’re serving. Not there for schema.org, rather verifiable claims<br/>
… specs as verifiable claims have a shared vocabulary, with some domains wide open, but people can’t change semantics<br/>
… in spec, intentionally or otherwise<br/>
Benjamin Young: chief aim of proposal is that json and json-ld world co-exist. Term sealing is to guarantee that …<br/>
… json and json-ld authors work on exact same shape<br/>
… maybe take a step back from sealing and see what other options there may be<br/>
… notion of foundational shape surviving framing<br/>
Ivan Herman: what I understand from ver. claim. is that they have a wrapper w/ properties whose value is an object…<br/>
… want to give total freedom for another vocabulary to leave locally. How about a syntax that says this is an extension point…<br/>
… @context null + specific situation that is sealing + @context null<br/>
… concentrate on this use case and try not to generalize<br/>
Rob Sanderson: sealing doesn’t solve actual shapes - just relationship between key and class. Not producing necessary structure<br/>
… it is intersection of RDF semantic layer and JSON keys. Not shape thing as an API. Mapping not API<br/>
Pierre-Antoine Champin: goal of sealing is not to guarantee LD has same shape as plain JSON, but JSON LD that has the same shape…<br/>
… has the same semantics.<br/>
… agree with ivan – sealed is sealed from point to rest of subtree and not try to address unsealed point.<br/>
David Newbury: Use case for unsealing is to insert user specified documents into structure. Unstructured JSON into LD…<br/>
Ivan Herman: may try to pick up json literal again<br/>
Ivan Herman: some examples don’t want literal, but piece of RDF graph but based on another vocabulary on its own<br/>
David Newbury: do we need ability to interpret unsealed part in context of whole document?<br/>
Gregg Kellogg: yes<br/>
David Newbury: do we have use cases where sealed and unsealed need to be interpreted?<br/>
Rob Sanderson: I think so<br/>
Rob Sanderson: Note - if we can’t solve by the end of the day, revisit the fallback position of JSON literals<br/>
Gregg Kellogg: We’re making this too hard. A sealed context is a collection of sealed terms – maintains original interpretation<br/>
… an extension point is where you use a term that is not sealed…<br/>
… say there is a named term in VC that has a given meaning, but want to be able to add other properties that aren’t sealed…<br/>
… also need a “this is a place for you to put stuff” – creating a scoped context where properties can be re-interpreted.<br/>
… sealed terms have fixed meaning when you use them.<br/>
Pierre-Antoine Champin: a term may be sealed (i.e. its meaning, esp. its IRI, can not be overridden) and an extension point (allow sealed terms to be overridden in its subtree)<br/>
David I. Lehn: common use will be use an example, preserving semantics and adding on…<br/>
Adam Soroka: we have ways to control shapes of json and RDF, but now we’re talking about a third way… controlling semantics<br/>
Pierre-Antoine Champin: gregg says extension points are properties that aren’t sealed — sealed means one cannot override<br/>
Gregg Kellogg: one approach is sealed context that allows context underneath to be unsealed…<br/>
Pierre-Antoine Champin: thanks @gkellogg, that’s clearer :)<br/>
Ivan Herman: i would like to hear a description of the exact sealing proposal<br/>
Gregg Kellogg: that would be pchampin’s PR<br/>
Ivan Herman: if we cannot explain where we are, we have a problem<br/>
Pierre-Antoine Champin: I agree with first 2 parts of the PR, the 3rd one is related to extension points<br/>
… we don’t have the same model for unsealing<br/>
… especially when it happens automatically<br/>
… I think the sealed definitions in a context cannot be overwritten by anything in that context<br/>
Ivan Herman: if you seal a context, then the only way the terms can be overridden is in a scoped context within the sealed context?<br/>
Pierre-Antoine Champin: I think so<br/>
Gregg Kellogg: explicates example on whiteboard Test so03<br/>
David I. Lehn: there are problems with this example because the inner context has no effect<br/>
Ivan Herman: we often use such examples, which I find a little unnatural<br/>
Harold Solbrig: That confuses me<br/>
Ivan Herman: [discusses example]<br/>
Gregg Kellogg, Ivan Herman, Rob Sanderson: [edit and discuss example]<br/>
Ivan Herman: sealing seems to mean that the terms are cast in concrete<br/>
Ivan Herman: it is very difficult to explain exception<br/>
David I. Lehn: Not sure why it’s unnatural?<br/>
Ivan Herman: if the context says “Sealed” that would seem to me to mean that everything is cast in concrete<br/>
… but that doesn’t seem to be the case<br/>
… we can change things with scoped contexts etc<br/>
Gregg Kellogg: this comes out of discussion of multiple sealed contexts<br/>
Harold Solbrig: by changing this one term to add a typo, I can open up everything I thought I was sealing<br/>
David Newbury: I would have assumed that by sealing a term, it would remain sealed<br/>
Gregg Kellogg: so once sealed, always sealed<br/>
Gregg Kellogg: that didn’t work for Verifiable Claims because they don’t want to intro a constraint on the JSON-LD within the wrapper<br/>
jeff mixer: can I constrain the type of value used with a term this way?<br/>
<b>all:</b> no, JSON-LD has no way to constrain the types of values<br/>
Adam Soroka: Agree completely. The wrapper compresses the protocol into the document<br/>
… It pulls us toward sealing other things beyond just the mapping, but into complex chains<br/>
Harold Solbrig: preventing people from accidentally overwriting contexts seems within our scope, but not all the further ramifications<br/>
Pierre-Antoine Champin: re: when terms become unsealed<br/>
… it is genuinely hard to explain<br/>
… I would say that a sealed term is sealed anywhere, except when the sealed context opens it<br/>
… there are two situations in which one can redefine a sealed term<br/>
… plain JSON users don’t expect the semantics to be sealed any more<br/>
… we should leave the decision of when to unseal terms to the authors of the sealed context<br/>
Ivan Herman: an extension point means a combination of “context now” and “I cannot touch what is on the upper level”<br/>
Gregg Kellogg: {discusses Ivan’s example]<br/>
Ivan Herman: how about a keyword “extension” that means you cannot redefine anything you inherit?<br/>
Gregg Kellogg: not quite @context: null.<br/>
David Newbury: if we have a context that defines ‘name’ within a sealed context, [points to examples]<br/>
Gregg Kellogg: [discusses examples]<br/>
… it may be simpler in the end to say that sealed terms are always sealed except for @context:null<br/>
… but subtleties arise when two sealed contexts interact<br/>
Rob Sanderson: can we do that example?<br/>
Ivan Herman: the only place where @context:null would have a nuclear option would be in the context that is sealed<br/>
Adam Soroka: [discussion of various examples]<br/>
Gregg Kellogg: [refers to examples]<br/>
David Newbury: ditto<br/>
Pierre-Antoine Champin: [example of two contexts conflicting]<br/>
David Newbury: the confusion caused by accounting for the complex interactions outweighs the utility<br/>
Harold Solbrig: what if we turned the pattern inside out and let contexts signify that they don’t intend to override?<br/>
… if you want to enforce that, you do it on a vocabulary-specific basis<br/>
… with rules for that particular context<br/>
Gregg Kellogg: so if a context lacked that flag, it would be harder to reuse it<br/>
Harold Solbrig: depends on how much you care about this relationship between the contexts<br/>
Adam Soroka: [discussion of what is or is not in scope for the use case and interoperability concern]<br/>
Gregg Kellogg: not clear how to specify that<br/>
Ivan Herman: going back to David’s example<br/>
Ivan Herman: [discusses example]<br/>
Rob Sanderson: [discusses example]<br/>
Gregg Kellogg: @context: null has a specific meaning and we are altering it<br/>
Ivan Herman: if we go down the road of “sealing”, then @context:null is going to conflict in some way and we will have to resolve it<br/>
Harold Solbrig: that’s why I went for an extension-based approach<br/>
Gregg Kellogg: we have until noon!<br/>
Harold Solbrig: can we get the use case in a more concrete form?<br/>
David Newbury: I’m increasingly fond of Harold’s idea to put in an “extension” property that errors if something with it overrides a term<br/>
… this allows me to write extensions that will work, but it may not be what VC need<br/>
Pierre-Antoine Champin: I have to think about that idea (Harold’s idea). gkellogg convinced me to try to explain again how a context becomes ineffective.<br/>
… I’m in favor of saying that whenever I traverse a term that is not part of a sealed context, it’s not sealed anymore<br/>
… after which a term could have been redefined: no guarantees about the semantics!<br/>
… as for @context:null– maybe the “extension” idea is a better way to make extension points; it makes the unsealing specific<br/>
… to make @context:null work one goes through both extension properties<br/>
Gregg Kellogg: let’s for the moment let’s put off the table any terms from a non-sealed context<br/>
… the behavior I heard dlongley describe<br/>
… is that history is wiped out and we start from scratch<br/>
Ivan Herman: I’m good so far<br/>
Gregg Kellogg: but if you do want to allow a mixture of sealed and unsealed terms, we get into the case that pchampin discussed<br/>
… two ways to have an extension point: 1) an unsealed term, under which anything goes, 2) a sealed term that is defined as an extension point<br/>
… and 2) doesn’t satisfy the use case<br/>
Ivan Herman: let’s say I have a claim, and I put into the context a @context:null to make an extension point.<br/>
… the danger is that I will redefine a term from VC, accidentally or maliciously<br/>
… IIUC VC, it’s a bit like Annotation<br/>
… outsiders are expected to use their own vocabularies with these guys, without having to understand VC or Annotation<br/>
Gregg Kellogg: one case is that everything is wiped out, but the other is that you intermingle the use of properties<br/>
Ivan Herman: so I can add my own context<br/>
Gregg Kellogg: but not redefine terms<br/>
… we would end up with semantics that differ between JSON and JSON-LD<br/>
David I. Lehn: the way that Activity Streams does this is to say “You must not override or change contexts” as a normative piece (see <a href="https://www.w3.org/TR/activitystreams-core/#jsonld)">https://www.w3.org/TR/activitystreams-core/#jsonld)</a><br/>
Rob Sanderson: people do this now by ordering the contexts<br/>
Rob Sanderson: [returns to examples]<br/>
Rob Sanderson: you can say that a sealed context can have an extension<br/>
… having sealed on the context is confusing and makes more sense at the term level<br/>
Ivan Herman: what is the difference between sealed and extension<br/>
… if extension was not true then there would be a warning<br/>
Ivan Herman: why not just say extension:true – it’s not setting a context?<br/>
Rob Sanderson: to distinguish from the case of this term being an extension<br/>
Gregg Kellogg: two different keywords that act on the context are unclear<br/>
Rob Sanderson: the change is that the default condition is “sealed”<br/>
Dave Longley: i’m on another call and can’t join, so i apologize… but was reading what’s here and thought i’d throw this in here… if there is a fear that schema.org (or contexts like it) might get sealed … we could allow {"@context": [{"@contextref": "<a href="http://schema.org",">http://schema.org",</a> "@sealed": false}, ...]} and the reason that would be ok for the sealed case we care about in VC community and so on, is because JSON processors would reject that … the context MUST be specified as a simple string, that’s what JSON-only processors would be looking for: {"@context": ["<a href="https://w3id.org/credentials/v1",">https://w3id.org/credentials/v1",</a> ...]}<br/>
David Newbury: I thought we were talking about three use cases, one of which is ability to delete context, which is @context:null<br/>
… are we trying to prevent people from overriding accidentally or maliciously?<br/>
Gregg Kellogg: we can’t prevent malice<br/>
Ivan Herman: it’s a clean slate… but not<br/>
Harold Solbrig: @context:null is often too much<br/>
Adam Soroka: gkellogg and ivan: [thrash more examples]<br/>
Rob Sanderson: we fundamentally don’t want the semantics to slip between JSON and JSON-lD<br/>
Rob Sanderson: [more examples]<br/>
Rob Sanderson: gkellogg: “extension” unseals everything, which is not @context:null<br/>
Adam Soroka: [discusion of @context:null and whether it is a problem for the original use case]<br/>
Gregg Kellogg: if there’s another context with sealed terms and extension points<br/>
Rob Sanderson: regardless of which context has the sealing, extension overrides it.<br/>
Gregg Kellogg: that’s the simplest way<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-442896360">https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-442896360</a><br/>
Benjamin Young: dlongley and I discuss term ordering and inheritance<br/>
… there is a problematic example there<br/>
… [discusses vocabulary]<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-454485321">https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-454485321</a><br/>
Benjamin Young: this example shows how to avoid that by being verbose<br/>
… [discusses problems that concern @type usage and inheritance]<br/>
David Newbury: if I have two sealed contexts that define the same term, who wins?<br/>
Gregg Kellogg: last one in<br/>
Rob Sanderson: Reviewing the rules that we think we’re talking about.<br/>
Adam Soroka: the use case is to avoid semantic slippage around terms when interpreted as JSON.<br/>
Rob Sanderson: If @extension:true is encountered in scoped context definition, then attempts to redefine terms from the active context will succeed.<br/>
Ivan Herman: If this is the rule, in the embedded context, I can also change the ID of that term?<br/>
Rob Sanderson: yes. You can redefine anything.<br/>
Ivan Herman: so if I have @extension : {}, the extension of the original key remains.<br/>
Gregg Kellogg: the effect of extensions: true is to update the active context and remove the sealed status of all terms within that context<br/>
… allowing those terms to be renamed within the value of the extension property<br/>
Ivan Herman: so inside, something may be different than outside?<br/>
Gregg Kellogg: yes. It only makes sense within a scoped context, and only makes sense on a sealed term.<br/>
… if you extend an unsealed property…<br/>
Gregg Kellogg: within an extension property, we can only define additional properties within that context.<br/>
Rob Sanderson: this is an error: [{"extension": { @id: extension_property, @sealed:true, @context: {@extension:true}}}}, { "extension":{@id: extension_propery}]<br/>
Rob Sanderson: extending unsealed terms is dumb, but not an error.<br/>
Pierre-Antoine Champin: I’m fine with this…in practice, this would mean that the plain JSON interpretation can only trust terms from the sealed context in the subtrees I defined.<br/>
… because when there is a term from another context, theydon’t know anything about that term.<br/>
… if we’re happy with this, we’re happy, but it seems like a lot of work to get to the same point.<br/>
Gregg Kellogg: I don’t think we’re at the same point.<br/>
… it’s only when we transition to a property when it’s extension: true that we transition<br/>
Pierre-Antoine Champin: the assumption that JSON-only is the same–it’s only trustworthy when looking at sealed terms.<br/>
Gregg Kellogg: that’s consistent with the spec.<br/>
Pierre-Antoine Champin: right. that’s why I’m OK with this.<br/>
Rob Sanderson: the thing we’re not able to deal with is nulling out terms and then redetermining them. So it works.<br/>
… so for VC, the context for the claim can start with null, which wipes out their context.<br/>
… so I don’t think that VC will use extension: true.<br/>
Gregg Kellogg: do we have a use case for this?<br/>
Ivan Herman: I thought that they want to avoid that anyone includes in the claim things that redefine things within verifiable claims<br/>
Gregg Kellogg: I think that applies to the envelope<br/>
… within the envelope, naive applications may not need to understand them.<br/>
… for example, annotation properties<br/>
… they come from another context<br/>
… so if you introduce another term in the envelope, those envelope extensions won’t override the VC<br/>
Rob Sanderson: don’t mess with our stuff<br/>
… vs:issuer should not be overridden, but if an extension defines ex:date, that’s fine. but the extension should not be able to override vs:issuer.<br/>
Gregg Kellogg: …explaining his example…<br/>
… since we have both a sealed context and schema, name remains it’s context. within the extension, something could set the scoped context to null.<br/>
… it could add a new context, and it could be anything.<br/>
… thus, VC does not need extension<br/>
Ivan Herman: why are we doing this?<br/>
… if this can be met with sealed and nullification, what is the use case for extension?<br/>
… this is the simplest thing that can work. That’s what we should do.<br/>
… do we need extension at any point?<br/>
Gregg Kellogg: I agree.<br/>
Adam Soroka: so there’s no need for nullification<br/>
… (i mean extension)<br/>
Gregg Kellogg: what we determined is that we’re not going to restrict @context: null. In that case, we don’t need the exceptions.<br/>
… you can always do @context: null<br/>
… thus, you can do it in a scoped context<br/>
… meeting VC’s use case<br/>
Benjamin Young: this VC is something that we haven’t discussed<br/>
… We don’t have anything that addresses it .<br/>
… since it’s type term usage, not property-term usage<br/>
… can we describe it as a use-case?<br/>
… term inheritance works…they don’t need sealing…<br/>
… but the scoped context around @type breaks things<br/>
… see the comment further down about getting around this by being verbose, but…this is how those things are extended.<br/>
… using a base type, and adding additional types…so type ordering matters.<br/>
… endorsed by handwaving, but would like a better mechanism<br/>
… but this is where things are breaking<br/>
Benjamin Young: example description is here <a href="https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-442896360">https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-442896360</a><br/>
Benjamin Young: verbosity “solution” is here <a href="https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-454485321">https://github.com/w3c/json-ld-syntax/issues/20#issuecomment-454485321</a><br/>
Gregg Kellogg: this example: we would process the scoped context for specificCredential…issue is defined in an example.<br/>
… this is dlongley’s example, not mine<br/>
… he was saying that the context ordering gets nullified by scope contexts and type…core vocabulary at the end doesn’t work, since types can override that.<br/>
Gregg Kellogg: scoped context would not override this if issuer is a sealed term<br/>
Benjamin Young: we have two inheritance orders: CSS and FIFO<br/>
… so we would flip them, so if sealing isn’t done properly, then we have problems<br/>
… don’t know if there’s a solution here…other than a core vocabulary that comes first<br/>
Rob Sanderson: this seems to me to be application-logic-layer<br/>
… so VC would have a sealed context…<br/>
Benjamin Young: {"@context": {"@sealed": "<a href="http://vc-data.example/",">http://vc-data.example/",</a> "@extensions": ["<a href="http://extension-1.exampe/",">http://extension-1.exampe/",</a> "<a href="http://other.example/extension"]}}">http://other.example/extension"]}}</a><br/>
Rob Sanderson: …typing on the whiteboard…<br/>
Gregg Kellogg: so there’s a new extension keyword that imports without redefining?<br/>
Benjamin Young: this is something totally different<br/>
Rob Sanderson: this just inverts where the properties live<br/>
Benjamin Young: this means that the core is at the top, and the extensions below<br/>
Rob Sanderson: with would work with term definitions<br/>
Benjamin Young: currently, you put the core last.<br/>
… now you need to move it to the top, or potentially to both places<br/>
Gregg Kellogg: this means that the thing at the front and seal it, then it can’t be changed<br/>
Rob Sanderson: to me, sealing is the inversion of the property rules.<br/>
Benjamin Young: this means that the scope documents now need to be explicitly 1.1<br/>
Rob Sanderson: if you’re using 1.1, but interpreting as 1.0, it will be strange.<br/>
Gregg Kellogg: I think we can only talk about what 1.1 will do<br/>
… the reason that 1.1 is needed, so that if a 1.0 processor sees it, it will die<br/>
Benjamin Young: so in 1.1 we’re inverting the order?<br/>
Gregg Kellogg: I don’t see it as inverting, I see it as a new effect.<br/>
Benjamin Young: this sounds like !important in CSS<br/>
Gregg Kellogg: it continues to work<br/>
Benjamin Young: even if someone else seals it first?<br/>
Gregg Kellogg: I think we need to table multiple contexts that seal<br/>
… I don’t think it’s what we’re promoting<br/>
… as a best practice, there should be only one<br/>
… as a corner case, we should…<br/>
Ivan Herman: the usage of sealed in the grand scale of things will be minimal<br/>
… if it’s awkward, then that’s OK, because this is rarely used.<br/>
Adam Soroka: we don’t know this will be rare<br/>
Ivan Herman: our schema discussion showed us that we should not use sealed regularly.<br/>
… common, vague vocabularies should not used sealed.<br/>
… there are good use cases, but let’s be careful.<br/>
Gregg Kellogg: IMO, sealed should only be used by contexts defined in specifications where there is a dual JSON/JSON-LD processing mode, and then only rairly<br/>
Rob Sanderson: the other thing that’s important to put down is sealed is not a guarantee that it can’t be changed, but to prevent unintentional.<br/>
… you can always put in a version 1.0…<br/>
Gregg Kellogg: you can’t go back<br/>
Ivan Herman: the @context:null option is there, so we can not defend against all malicious use<br/>
Rob Sanderson: so we don’t need extension: true, because there’s no use case and we can cover it by nulling and re-i porting?<br/>
… so are we back to the simple rule?<br/>
… If @sealed: true is encountered on a term definition, then attempts to redefine the term will fail and result in a warning.<br/>
Gregg Kellogg: with a note about using context:null<br/>
Adam Soroka: so there’s no way that a term picks up an unintentional meaning<br/>
Harold Solbrig: can I put a @sealed on the external context?<br/>
Gregg Kellogg: you can’t do that<br/>
Ivan Herman: you can’t seal someone else’s context<br/>
Harold Solbrig: I would still like to be able to make sure that I don’t unintentionally reuse terms<br/>
Gregg Kellogg: can this be solved with SHEX?<br/>
Harold Solbrig: maybe…<br/>
Pierre-Antoine Champin: what happened to redefining a term within a sealed context?<br/>
Gregg Kellogg: we said that’s not possible.<br/>
… the only way to do this is to null the context<br/>
Pierre-Antoine Champin: so context:null could happen anywhere?<br/>
Rob Sanderson: correct<br/>
Pierre-Antoine Champin: so sealing prevents explicit redefining the sealed terms, but not guaranteeing their stability.<br/>
Gregg Kellogg: correct<br/>
Rob Sanderson: context: null is the way out of things, and you’d only do that in the VC case if you were trying to be malicious, and we’ve decided we can’t guard against that.<br/>
… the intent is to avoid unintentional redefining.<br/>
… to fix that, you need a preprocessor.<br/>
Adam Soroka: is sealing now to strong a name?<br/>
Gregg Kellogg: it is sealed in the active context.<br/>
Pierre-Antoine Champin: I disagree; freezing (and sealing, for that matter) in JS is a quite different thing<br/>
Ivan Herman: let’s make a resolution, and go to lunch<br/>
Proposed resolution: Adopt the following rules for sealing - 1: If @sealed:true is encountered on a term definition in the active context, then attempts to redefine the term will fail and result in a warning. 2 - If @sealed:true is encountered on a context, then all terms defined within the context should be treated as having @sealed:true in their definition. 3- If @sealed:false is encountered on a term definition in the active context, where the context has been sealed with @sealed:true, then the sealing for that term is removed. (Rob Sanderson)<br/>
Harold Solbrig: if there’s a sealed: false for an unsealed term, what happens?<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: nothing–that doesn’t mean anything<br/>
Adam Soroka: should we warn?<br/>
Gregg Kellogg: +1<br/>
Jeff Mixter: +1<br/>
David Newbury: +1<br/>
Simon Steyskal: +1<br/>
Harold Solbrig: +1<br/>
Pierre-Antoine Champin: +0.5<br/>
Gregg Kellogg: we SHOULD<br/>
Adam Soroka: +1<br/>
Benjamin Young: +0 (would like to see use case mappings for all these things)<br/>
David I. Lehn: +1<br/>
Resolution #1: Adopt the following rules for sealing - 1: If @sealed:true is encountered on a term definition in the active context, then attempts to redefine the term will fail and result in a warning. 2 - If @sealed:true is encountered on a context, then all terms defined within the context should be treated as having @sealed:true in their definition. 3- If @sealed:false is encountered on a term definition in the active context, where the context has been sealed with @sealed:true, then the sealing for that term is removed. {: #resolution1 .resolution}<br/>
Dave Longley: “If @sealed:false is encountered on a term definition in the active context, where the context has been sealed with @sealed:true, then the sealing for that term is removed.” &lt;– this sounds like a big problem as written …<br/>
Dave Longley: is this saying that with {"@context": [1, 2]} that 2 can unseal anything it wants?<br/>
Dave Longley: if so, that defeats the purpose of sealing … so i can only imagine that something else was meant by that text.<br/>
Dave Longley: it sounds like an attacker that defines 2 can get a VC to be successfully processed by a JSON-only processor in a different way from a JSON-LD processor under that condition<br/>
Dave Longley: i’ll check back here later to see if there’s a clarification<br/>
Dave Longley: if the ability to unseal a sealed context is a requirement (as opposed to just extending in an acceptable way) … it seems like we could do it without messing up JSON-only processors by using some other syntax that they would otherwise reject.<br/>
Dave Longley: JSON-only processors will be looking for {"@context": [<specific string>, <specific string>, ...ignored]}<br/>
Dave Longley: or could perhaps be further limited to: {"@context": [<specific string>, <specific string>, ...<must be a string but value is ignored>]}<br/>
Dave Longley: in other words, any deviation from some very rigid rules would cause a JSON-only processor to reject the input … which could be potentially used to our advantage if we need to cover other use cases where we are fearful of people sealing things that JSON-LD aware entities would like to unseal in a way that would not impact JSON-only processors.<br/>
Dave Longley: but what is key is that that any mechanism that is used to unseal would not impact JSON-only processors.<br/>
Dave Longley: if we can abide by that rule, i don’t see why i wouldn’t be supportive.<br/>
Rob Sanderson: dlongley: The @sealed:false would only work for the original definition, no subsequent attempt to redefine it<br/>
Rob Sanderson: It overrides the @sealed:true on the context<br/>
Rob Sanderson: But not @sealed:true on a term<br/>
Dave Longley: azaroth: ok, if I understand you I think that’s fine<br/>
Adam Soroka: <a href="https://github.com/w3c/json-ld-syntax/issues/87">https://github.com/w3c/json-ld-syntax/issues/87</a><br/>
2.1. closing related issues<br/>
Proposed resolution: Close #87, as unnecessary given above rules. (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
David Newbury: +1<br/>
Jeff Mixter: +1<br/>
Simon Steyskal: +1<br/>
David I. Lehn: +1<br/>
Resolution #2: Close #87, as unnecessary given above rules. {: #resolution2 .resolution}<br/>
Simon Steyskal: <a href="https://github.com/w3c/json-ld-syntax/issues/98">https://github.com/w3c/json-ld-syntax/issues/98</a><br/>
Proposed resolution: close #98, as unnecessary given the rules above (Ivan Herman)<br/>
David Newbury: +1<br/>
Ivan Herman: +1<br/>
Proposed resolution: Close #98, as @context: null clears context definitions (Rob Sanderson)<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Harold Solbrig: +1<br/>
David I. Lehn: +1<br/>
Resolution #3: Close #98, as @context: null clears context definitions {: #resolution3 .resolution}<br/>
Resolution #4: Close #98, as @context: null clears context definitions {: #resolution4 .resolution}<br/>
David I. Lehn: <a href="https://github.com/w3c/json-ld-syntax/issues/116">https://github.com/w3c/json-ld-syntax/issues/116</a><br/>
David I. Lehn: discussion of [<a href="https://github.com/w3c/json-ld-syntax/issues/116#issuecomment-460681505]">https://github.com/w3c/json-ld-syntax/issues/116#issuecomment-460681505]</a><br/>
David I. Lehn: [missed scribing much back and forth, trying to catch up on the gist]<br/>
Gregg Kellogg: [pointing out current algorithm could be modified based on some rules to handle the partial redefinition]<br/>
Ivan Herman: [concerns about backwards compatibility]<br/>
Gregg Kellogg: based on algorithm, 1.0 would fall through to new 1.1 behavior<br/>
Ivan Herman: sounds like adding more hacks<br/>
Dave Longley: also want it to be easier for json authors<br/>
Gregg Kellogg: we go towards side of being easier for authors<br/>
David Newbury: hard to follow what is going on without knowing all the processing steps<br/>
Gregg Kellogg: not overriding term itself […?]<br/>
Rob Sanderson: more authors to explain to than people needing to understand this feature. on side of authors.<br/>
David Newbury: issue with how things are overridden. seeing next term definition would think it fully overrode first one.<br/>
Ivan Herman: [worries about more complications]<br/>
David Newbury: example seems like json in json-ld<br/>
David I. Lehn: it’s a mix in this case, using json tools to process json-ld. would want to know the context of the data.<br/>
Rob Sanderson: [discussing how some examples would be processed]<br/>
Ivan Herman: adding new features to already complex system. really worried about it.<br/>
Dave Longley: understand ivan. A goal is to make ergonomics of 1.1 easier for users and better than 1.0<br/>
Proposed resolution: if @id is not present in a term definition, and there is no default vocabulary, but has an expansion to an absolute IRI in the active context, then the @id of the new definition is taken from the active context (otherwise this is still an error) (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Simon Steyskal: +1<br/>
Gregg Kellogg: +1<br/>
Dave Longley: +1<br/>
David I. Lehn: +1<br/>
Jeff Mixter: +1<br/>
David Newbury: +0.5<br/>
Adam Soroka: +1<br/>
Resolution #5: if @id is not present in a term definition, and there is no default vocabulary, but has an expansion to an absolute IRI in the active context, then the @id of the new definition is taken from the active context (otherwise this is still an error) {: #resolution5 .resolution}<br/>
Proposed resolution: Once previous is done, we can close #116 (Rob Sanderson)<br/>
Harold Solbrig: +1<br/>
Ivan Herman: +1<br/>
Rob Sanderson: +1<br/>
David Newbury: +1<br/>
Simon Steyskal: +1<br/>
Jeff Mixter: +1<br/>
David Newbury: +1<br/>
Rob Sanderson: +1<br/>
Harold Solbrig: +1<br/>
Adam Soroka: +1<br/>
Ivan Herman: +1<br/>
Rob Sanderson: Really 116 :)<br/>
David I. Lehn: +1<br/>
Resolution #6: Once previous is done, we can close #116 {: #resolution6 .resolution}<br/>
Proposed resolution: We agree with the processing order per #61: scoped definition from property, then from type, then embedded. PR to update docs to come. (Rob Sanderson)<br/>
Ivan Herman: +1<br/>
Rob Sanderson: +1<br/>
Adam Soroka: +1<br/>
David I. Lehn: +1<br/>
Simon Steyskal: +1<br/>
David Newbury: +1<br/>
Jeff Mixter: +1<br/>
Harold Solbrig: +1<br/>
Gregg Kellogg: +1<br/>
Resolution #7: We agree with the processing order per #61: scoped definition from property, then from type, then embedded. PR to update docs to come. {: #resolution7 .resolution}<br/>
Harold Solbrig: scribnic hsolbrig<br/>
Gregg Kellogg: what if you said context null?<br/>
Rob Sanderson: it would wipe out all definitions<br/>
Gregg Kellogg: you could also seal activities<br/>
</details>
